### PR TITLE
Adding SAS Credential Support to the Event Grid SDK

### LINKED ISSUE #13166: Add SAS credential support to the SDK
We will need a new type (maybe SharedAccessSignatureCredential) to support SAS token credentials. It may be that we want this in Azure.Core, but for now we can start off by putting it in Event Grid. We should add a constructor overload that has this new type as a parameter.

There is an ongoing discussion as to the best place to put SAS token generation - client vs a separate static class, but we seem to be in agreement that this should be included.

Let's include it in the client for now and add some unit tests and live tests involving the SAS token. The unit tests would verify that the token is being generated in the expected format. The live test would verify that we are actually able to authenticate with the service using the token.

### PR DESCRIPTION
**Pull request by kerri-lee**
Added a new SharedAccessSignatureCredential type to the SDK and a method to generate a SAS token in the client. Also added a live test using the token to publish events.

### CODE REVIEW

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs [hunk 1]
```diff
@@ -35,20 +41,38 @@ public EventGridClient(Uri endpoint, AzureKeyCredential credential)
...
             _clientDiagnostics = new ClientDiagnostics(options);
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridClient"/> class.
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <param name="credential"></param>
+        /// <param name="options"></param>
+        public EventGridClient(Uri endpoint, SharedAccessSignatureCredential credential, EventGridClientOptions options = default)
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs:66]: nit: we should follow the same pattern as for AzureKeyCredential ctor where we have separate ctors, one having options and one not having it.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs [hunk 2]
```diff
@@ -163,5 +187,39 @@ public virtual Response PublishCustomEvents(IEnumerable<object> events, Cancella
...
+        /// <returns></returns>
+        public string BuildSharedAccessSignature(DateTimeOffset expirationUtc)
+        {
+            const char Resource = 'r';
+            const char Expiration = 'e';
+            const char Signature = 's';
+
+            if (_key == null)
+            {
+                throw new NotSupportedException();
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs:204]: nit: we should include a message in the exception that explains this method can only be used when using an EventGridClient created using AzureKeyCredential.

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs [hunk 3]
```diff
@@ -58,6 +58,25 @@ public async Task CanPublishCloudEvent()
             await client.PublishCloudEventsAsync(GetCloudEventsList());
         }
 
+        [Test]
+        public async Task CanPublishEventUsingSAS()
+        {
+            EventGridClientOptions options = Recording.InstrumentClientOptions(new EventGridClientOptions());
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs:64]: nit: move the options variable down to where it is used or just add it inline into the ctor.

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs [hunk 4]
```diff
@@ -58,6 +58,25 @@ public async Task CanPublishCloudEvent()
...
+            EventGridClient client = new EventGridClient(
+                new Uri(TestEnvironment.TopicHost),
+                new AzureKeyCredential(TestEnvironment.TopicKey));
+
+            string sasToken = client.BuildSharedAccessSignature(DateTimeOffset.UtcNow.AddMinutes(60));
+            TestContext.Progress.WriteLine(sasToken);
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs:70]: nit: remove

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs:1]: The pattern we use in EH/SB is to call the Live test classes something like EventGridClientLiveTests, and the ones with unit tests EventGridClientTests. Would it make sense to do that here?

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+namespace Azure.Messaging.EventGrid.Tests
+{
+    public class EventGridClientUnitTests : ClientTestBase
+    {
+        public EventGridClientUnitTests(bool async) : base(async)
+        {
+        }
+
+        [Test]
+        public void MustSetKeyCredential()
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs:19]: nit: typically we name tests that are asserting an exception is thrown something like BuildSharedAccessSignatureThrowsWhenCalledFromSasClient

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Azure.Core.TestFramework;
+using NUnit.Framework;
+
+namespace Azure.Messaging.EventGrid.Tests
+{
+    public class EventGridClientUnitTests : ClientTestBase
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientUnitTests.cs:12]: I don't think this needs to inherit from ClientTestBase.

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs [hunk 8]
```diff
@@ -3,10 +3,13 @@
 
 using System;
 using System.Collections.Generic;
+using System.Globalization;
 using System.Net;
+using System.Security.Cryptography;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Web;
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs:12]: Are we using this?

Comment (kerri-lee) [reply sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs:12]: Yes the BuildSharedAccessSignature method uses the HttpUtility class

File: sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs [hunk 9]
```diff
@@ -2,156 +2,29 @@
...
-                    subject = $"Subject-{i}",
-                    topic = $"Topic-{i}"
-                });
-            }
+            EventGridClient sasTokenClient = new EventGridClient(
+                new Uri("https://exampletopic.westus2-1.eventgrid.azure.net/api/events"),
+                new SharedAccessSignatureCredential("thisIsNotAFakeCredential"));
 
-            return eventsList;
+            Assert.That(() => sasTokenClient.BuildSharedAccessSignature(DateTimeOffset.UtcNow.AddMinutes(60)),
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/tests/EventGridClientTests.cs:26]: nit: parameter indendation - use new line for each param within BuildSharedAccessSig

File: sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs [hunk 10]
```diff
@@ -163,5 +195,39 @@ public virtual Response PublishCustomEvents(IEnumerable<object> events, Cancella
                 throw;
             }
         }
+
+        /// <summary>
+        /// Creates a SAS token for use with Event Grid service
+        /// </summary>
+        /// <param name="expirationUtc">Expiration time</param>
+        /// <returns></returns>
```
Comment (JoshLove-msft) [inline sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridClient.cs:203]: nit: can we fill out the returns value and maybe explain a bit more about what the expiration time means based on the service documentation?

**Review by JoshLove-msft — APPROVED** [review]
Looks great!

**Pull request merged**